mountpoint-s3-crt/src/s3/endpoint_resolver.rs

```diff
@@ -0,0 +1,265 @@
+        let inner = unsafe { aws_s3_endpoint_resolver_new(allocator.inner.as_ptr()).ok_or_last_error()? };
+        Ok(Self { inner })
+    }
+
+    /// Resolve the endpoint with given Request Context and ruleset.
+    pub fn resolve(&self, context: RequestContext) -> Result<ResolvedEndpoint, Error> {
+        // SAFETY: `aws_endpoints_rule_engine_resolve` ensures that it returns a non null `aws_endpoints_resolved_endpoint`
+        // or it will return an error. So, the out_endpoint is valid and non null.
+        unsafe {
+            let mut out_endpoint: *mut aws_endpoints_resolved_endpoint = ptr::null_mut();
+            aws_endpoints_rule_engine_resolve(self.inner.as_ptr(), context.inner.as_ptr(), &mut out_endpoint)
+                .ok_or_last_error()?;
+            match aws_endpoints_resolved_endpoint_get_type(out_endpoint) {
+                aws_endpoints_resolved_endpoint_type::AWS_ENDPOINTS_RESOLVED_ERROR => Err(Error::last_error()),
+                _ => Ok(ResolvedEndpoint {
+                    inner: NonNull::new_unchecked(out_endpoint),
+                }),
+            }
+        }
+    }
+}
+
+impl Drop for RuleEngine {
+    fn drop(&mut self) {
+        // SAFETY: `self.inner` is a valid `aws_endpoints_rule_engine`, and on Drop it's safe to decrement
+        // the reference count since we won't use it again through `self.`
+        unsafe {
+            aws_endpoints_rule_engine_release(self.inner.as_ptr());
+        }
+    }
+}
+
+/// Add the context to build the endpoint
+#[derive(Debug)]
+pub struct RequestContext {
+    inner: NonNull<aws_endpoints_request_context>,
+}
+
+impl RequestContext {
+    /// Creates a new endpoint Request Context.
+    pub fn new(allocator: &Allocator) -> Result<Self, Error> {
+        s3_library_init(allocator);
+        // SAFETY: `allocator.inner` is a valid aws_allocator and and we check the return is non-null.
+        // SAFETY: aws_endpoints_request_context_new will acquire a reference to keep it alive after this function call, so it's safe to return an owned version to it.
+        let inner = unsafe { aws_endpoints_request_context_new(allocator.inner.as_ptr()).ok_or_last_error()? };
+        Ok(Self { inner })
+    }
+
+    /// Add the parameter to request context whose value is in form of string
+    pub fn add_string(&mut self, allocator: &Allocator, name: &OsStr, value: &OsStr) -> Result<(), Error> {
+        // SAFETY: allocator.inner and self.inner should be valid pointers.
+        // `name` and `value` should be valid aws byte cursor not be modified further.
+        unsafe {
+            aws_endpoints_request_context_add_string(
+                allocator.inner.as_ptr(),
+                self.inner.as_ptr(),
+                name.as_aws_byte_cursor(),
+                value.as_aws_byte_cursor(),
+            )
+            .ok_or_last_error()
+        }
+    }
```

Comment:
In CRT, they were using the same allocator for request context, rule engine and resolved endpoint. So, I decided to pass them as argument. If we dont need to pass the same allocator, it will be much clearer and easier.